bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/BuildManager.java

```diff
@@ -740,14 +740,47 @@ public void endTopLevel(boolean needsBuild) {
 		autoBuildJob.build(needsBuild);
 	}
 
+	/**
+	 * Waits till autobuild finished. Tries to finish it as soon as possible.
+	 */
 	public void waitForAutoBuild() {
-		autoBuildJob.wakeUp(0);
-		try {
-			Job.getJobManager().join(ResourcesPlugin.FAMILY_AUTO_BUILD, null);
-		} catch (OperationCanceledException | InterruptedException e) {
-			// ignore
+		while (!(autoBuildJob.getState() == Job.NONE)) {
+			while (!(autoBuildJob.getState() == Job.RUNNING || autoBuildJob.getState() == Job.NONE)) {
```

Comment:
What does the while condition here and above wants to achieve? 
Why do we care about states if the `wakeUp()` itself checks for `Job.SLEEPING`?

Reply:
Please add comment so this is clear while reading this code without looking into other files